src/track/beats.cpp

```diff
@@ -1,29 +1,526 @@
+    }
+
+    if (n > 0) {
+        for (; next_beat != m_beats.end(); ++next_beat) {
+            if (n == 1) {
+                // Return a sample offset
+                return next_beat->getFramePosition();
+            }
+            --n;
+        }
+    } else if (n < 0 && previous_beat != m_beats.end()) {
+        for (; true; --previous_beat) {
+            if (n == -1) {
+                // Return a sample offset
+                return previous_beat->getFramePosition();
+            }
+            ++n;
+
+            // Don't step before the start of the list.
+            if (previous_beat == m_beats.begin()) {
+                break;
+            }
+        }
+    }
+    return kInvalidFramePos;
+}
+Bpm BeatsInternal::getBpm() const {
+    if (!isValid()) {
+        return Bpm();
+    }
+    return m_bpm;
+}
+bool BeatsInternal::isValid() const {
+    return m_iSampleRate > 0 && !m_beats.empty();
+}
+void BeatsInternal::setSampleRate(int sampleRate) {
+    m_iSampleRate = sampleRate;
+    updateBpm();
+}
+BeatsInternal::BeatsInternal()
+        : m_iSampleRate(0), m_dDurationSeconds(0) {
+}
+
+void BeatsInternal::initWithProtobuf(const QByteArray& byteArray) {
+    track::io::Beats beatsProto;
+    if (!beatsProto.ParseFromArray(byteArray.constData(), byteArray.size())) {
+        qDebug() << "ERROR: Could not parse Beats from QByteArray of size"
+                 << byteArray.size();
+    }
+    m_beatsProto = beatsProto;
+    generateBeatsFromMarkers();
+}
+
+void BeatsInternal::initWithAnalyzer(const QVector<FramePos>& beats,
+        const QVector<track::io::TimeSignatureMarker>& timeSignatureMarkers) {
+    m_beatsProto.set_first_frame_position(beats.at(0).getValue());
+    int bpmMarkerBeatIndex = 0;
+    for (int i = 1; i < beats.size(); ++i) {
+        VERIFY_OR_DEBUG_ASSERT(
+                beats.at(i) > beats.at(i - 1) && beats.at(i) >= FramePos(0)) {
+            qDebug() << "Beats not in increasing order or negative, discarding "
+                        "beat"
+                     << beats.at(i);
+        }
+        else {
+            FrameDiff_t beatLength = beats.at(i) - beats.at(i - 1);
+            Bpm immediateBpm(kSecondsPerMinute * m_iSampleRate / beatLength);
+            if (m_beatsProto.bpm_markers().empty() ||
+                    m_beatsProto.bpm_markers().rbegin()->bpm() !=
+                            immediateBpm.getValue()) {
+                track::io::BpmMarker bpmMarker;
+                bpmMarker.set_beat_index(bpmMarkerBeatIndex);
+                bpmMarker.set_bpm(immediateBpm.getValue());
+                m_beatsProto.add_bpm_markers()->CopyFrom(bpmMarker);
+            }
+            bpmMarkerBeatIndex++;
+        }
+    }
+
+    for (const auto& timeSignatureMarker : timeSignatureMarkers) {
+        m_beatsProto.add_time_signature_markers()->CopyFrom(
+                timeSignatureMarker);
+    }
+
+    VERIFY_OR_DEBUG_ASSERT(m_beatsProto.time_signature_markers_size() > 0) {
```

Comment:
Even after deleting my Mixxx directory, I'm hitting this debug assertion when I load a track onto a deck and the analysis finishes:
```
DEBUG ASSERT: "m_beatsProto.time_signature_markers_size() > 0" in function void mixxx::BeatsInternal::initWithAnalyzer(const QVector<mixxx::FramePos>&, const QVector<mixxx::track::io::TimeSignatureMarker>&) at /home/jan/Projects/mixxx/src/track/beats.cpp:348
Aborted (core dumped)
```